### DISCUSSION TITLE
`pkgdown` in Quarto?

### DISCUSSION THREAD

**Discussion by mcanouil** [quarto-dev/quarto-cli#1591] [Q&A] [answered]
With how easy it is to add pages to website and/or combine pages as a website, I was wondering about the possibility to include within a Quarto website one or several pkgdown-like documentation website.
Is a "translation" of `pkgdown` to the Quarto framework planned? (Kind of related to the multi-repo website #1282 in the idea I have in mind)

**Comment by cderv** [comment, marked as answer]
Here is a project by @edgararuiz to create package documentation in a quarto website with the same idea as pkgdown https://github.com/edgararuiz/ecodown

I believe sparklyr doc is a live version of it 

https://spark.rstudio.com/

This seems like a good existing solution for those looking to use quarto feature for their doc website 

Regarding including several package doc in same website would be only extending what has been already done.

**Reply by mcanouil** [reply]
Thanks for `ecodown`!

Should I assume nothing native/internal is on the project board of Quarto? To make it explicit in Python there is the "docstring" and `pydoc` module which kind of does what `roxygen2` does. So having Quarto able to generate both website documentation seems a nice feature to me (I guess it is possible to do the same for observable and Julia).

**Comment by johnkerl** [comment]
Hi @cderv -- I'm trying to do some proof-of-concept work for rendering a package's API docs and examples/vignettes in Quarto.

The latter works just fine.

The former, after several hours over several days, I confess I am still struggling even to conceptualize. It's clear that _if_ I have a `.Rmd` file I can go many places with it. That works. The hole I'm trying to fill is when my starting point is _not_ hand-written `.Rmd`. Rather, the starting point is all the docstrings within `R/*.R` in my package's source code. What `pkgdown` does is map `R/*.R` docstrings to `.html` files. What I want is something -- anything -- that will, similarly, map `R/*.R` docstrings into any file format Quarto can handle (`.Rmd`, `.qmd`, etc.).

Cloning the `sparklyr` repo, as of today anyway, I couldn't find any way in which it is using `ecodown`. I'd love it if `sparklyr` were a usable example, but I can't find any way in which `sparklyr` is actually using `ecodown` -- I simply can't find anything to copy/imitate.

I've also poked with `rmarkdown` and `ecodown` and with neither of them have I found a way to generate markdown files _starting from_ the docstrings in my package's `R/*.R` files. I acknowledge may be missing something obvious.

Does anyone here have any examples, however rudimentary or fragmentary, on approaches to solving this problem?

**Reply by cderv** [reply]
> Cloning the sparklyr repo, as of today anyway, I couldn't find any way in which it is using ecodown. I'd love it if sparklyr were a usable example, but I can't find any way in which sparklyr is actually using ecodown -- I simply can't find anything to copy/imitate.

Spark website lives at https://github.com/rstudio/spark.rstudio.com and the website is built using **ecodown** (https://github.com/rstudio/spark.rstudio.com/blob/main/.github/workflows/ecodown.yaml). So this would be your example. 

**ecodown** does not seem to be a package meant for external use, and more targeted for internal use. Though it is public so you can look at how `ecodown::ecodown_build()` works. 

This would be your example. 

> I've also poked with rmarkdown and ecodown and with neither of them have I found a way to generate markdown files starting from the docstrings in my package's R/*.R files. I acknowledge may be missing something obvious.

I don't know what **ecodown** is really doing, but I think building on **pkgdown** feature would probably be the best. Starting point would probably be .Rd file and not .R file. based on https://pkgdown.r-lib.org/reference/rd2html.html probably which uses [`tools::parse_Rd()`](https://developer.r-project.org/parseRd.pdf). I would try build on that. You could also try `rd2html()` for Rd to HTML, then convert to .md using Pandoc maybe (using **rmarkdown** or **pandoc**). 

Overall, you probably just need to convert to .md and not .qmd. 

I don't know any other tool parser that do .R from roxygen comment to .HTML or .md but you could probably create one.

**Reply by johnkerl** [reply]
Thanks @cderv !

**Reply by eitsupi** [reply]
You may interested in etiennebacher/altdoc#39 and etiennebacher/altdoc#46

**Reply by cderv** [reply]
Thanks for sharing this @eitsupi  !

**Comment by chendaniely** [comment]
This post shows up when searching for quarto + pkgdown. Since Python was mentioned in the accepted solution. there is a quarto <> python package documentation connection in `quartodoc`: <https://machow.github.io/quartodoc/get-started/overview.html>

**Reply by cderv** [reply]
There is also this new project in python world: https://github.com/rich-iannone/great-docs

**Comment by edgararuiz** [comment]
Hi all, I just realized that I haven't shared `ecodown`'s replacement, `pkgsite`. I intend to release this one to CRAN at some point this quarter. It is very focused on just processing package help files and nothing else. So it is more like `quartodoc` than it is about `pkgdown`. It is what makes it possible for me to have a single site for the Python and R versions of `mall`: https://mlverse.github.io/mall/reference/

`pkgsite`'s website is fully built in Quarto: https://edgararuiz.github.io/pkgsite/, which gives me dark mode for free

**Comment by eitsupi** [comment]
I'm not sure if this will meet your needs, but I've released a Rust tool that converts Rd files directly to qmd files.
It has a feature that mimics downlit processing, like pkgdown does internally, and inserts links to R reference sites as Markdown links.
https://github.com/eitsupi/rd2qmd

I haven't tested it thoroughly, but I hope it will be useful, as there are few solutions that convert Rd files directly to (Quarto) Markdown.
